// dao/MessageProvider.js
/**
 *
 * The MessageProvider class stores header messages for the application.
 * It also exposes a few methods to grab a random message to display for
 * each new message created.
 *
 */

MessageProvider = function() {};

MessageProvider.prototype.messages =
[
	'Filthy Sandraker.',
	'Prepare yourself!',
	'Victory will soon be ours!',
	'Prepare for deployment!',
	'No mercy this day.',
	'No quarter given.',
	'I am the greatest!'
];

MessageProvider.prototype.loadingMessages =
[
	'please wait while i buffer...',
	'your request is buffering itself... do not get angry...',
	'buffering... please make sure to like, fav and retweet on myspace...',
	'the requested audio clips are buffering... go grab a beer!',
	'B-B-B-B-B-B-UUUFFFEEERRRRRRIIIIIIIINNNGGGGGGGG!!!!!111!111!1'
];

/**
 * Grab a random message from the messages collection.
 *  - Used in the large header callout
 */
MessageProvider.prototype.getMessage = function()
{
	var rnd = Math.floor(Math.random() * (this.messages.length));
	return this.messages[rnd];
};

/**
 * Grab a random loading message from the loadingMessages collection.
 *  - Used on the initial loading screen.
 */
MessageProvider.prototype.getLoadingMessage = function()
{
	var rnd = Math.floor(Math.random() * (this.loadingMessages.length));
	return this.loadingMessages[rnd];
};

exports.MessageProvider = MessageProvider;
